File: identicon/src/main/java/com/lelloman/identicon/github/GithubIdenticonDrawable.java
package com.lelloman.identicon.github;


import android.graphics.Canvas;
import android.graphics.Paint;

import com.lelloman.identicon.IdenticonDrawable;


/**
 * 	Draws an identicon as the ones from github, it just uses 15bits
 * 	for the tiles and 6bits - 4 values for the colors
 */
public class GithubIdenticonDrawable extends IdenticonDrawable {

	// create 4*4*4 colors but remove the grays
	public static int[] COLORS;
	static {
		int n = 4;
		int size = n*n*n - n;
		COLORS = new int[size];
		int index = 0;
		int colorBase = 150;
		int colorTot = 100 / n;
		for(int i=0;i<n;i++){
			for(int j=0;j<n;j++){
				for(int k=0;k<n;k++){
					if(i == j && j == k) continue;
					int r = colorBase + i * colorTot;
					int g = colorBase + j * colorTot;
					int b = colorBase + k * colorTot;

					int color = 0xff000000 + r *0x10000 + g *0x100 + b;
					COLORS[index++] = color;
				}
			}
		}
	}

	private Paint mFgPaint = new Paint(Paint.ANTI_ALIAS_FLAG);

	public GithubIdenticonDrawable(int width, int height, int hash){
		super(width,height,hash,5);

		mFgPaint.setStyle(Paint.Style.FILL);
		invalidateBitmap();
	}

	public int getColor(){
		int i = Math.abs((getHash() >> 15) % COLORS.length);

		return COLORS[i];
	}

	@Override
	protected void drawBitmap(Canvas canvas) {

		mFgPaint.setColor(getColor());

		int w = (int) (canvas.getWidth() / 5f);
		int h = (int) (canvas.getHeight() / 5f);

		canvas.drawColor(0xffffffff);

		int i = 0;
		for(int x=0;x<3;x++){
			int startX = x*w;
			int endX = startX +w;

			int startXMirror = (4-x) * w;
			int endXMirror = startXMirror + w;

			for(int y=0;y<5;y++){
				int value = getHash() >> i++;
				boolean draw = value % 2 == 0;
				if(!draw) continue;

				float startY = y*h;
				float endY = startY+h;

				canvas.drawRect(startX,startY,endX,endY, mFgPaint);
				if(x != 2)
					canvas.drawRect(startXMirror,startY,endXMirror, endY, mFgPaint);
			}
		}
	}

}

File: identicon/src/main/java/com/lelloman/identicon/drawable/GithubIdenticonDrawable.kt
package com.lelloman.identicon.drawable

import android.graphics.Canvas
import android.graphics.Paint

import com.lelloman.identicon.util.extractBits

/**
 * Draws an identicon as the ones from github, it uses 15 bits
 * for the tiles and 18 bits (6+6+6) for the RGB color.
 */
class GithubIdenticonDrawable(
    width: Int,
    height: Int,
    hash: ByteArray
) : IdenticonDrawable(
    desiredWidth = width,
    desiredHeight = height,
    hash = hash,
    multipleOf = 5
) {

    private val fgPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply { style = Paint.Style.FILL }

    val color: Int
        get() {
            val r = 80 + extractBits(hash, 15, 6) * 2
            val g = 80 + extractBits(hash, 21, 6) * 2
            val b = 80 + extractBits(hash, 27, 6) * 2
            return -0x1000000 + r * 0x10000 + g * 0x100 + b
        }

    init {
        invalidateBitmap()
    }

    override fun drawBitmap(canvas: Canvas) {

        fgPaint.color = color

        val w = (canvas.width / 5f).toInt()
        val h = (canvas.height / 5f).toInt()

        canvas.drawColor(-0x1)

        var i = 0
        for (x in 0..2) {
            val startX = x * w
            val endX = startX + w

            val startXMirror = (4 - x) * w
            val endXMirror = startXMirror + w

            for (y in 0..4) {
                val draw = extractBits(hash, i++, 1) == 0
                if (!draw) continue

                val startY = (y * h).toFloat()
                val endY = startY + h

                canvas.drawRect(startX.toFloat(), startY, endX.toFloat(), endY, fgPaint)
                if (x != 2)
                    canvas.drawRect(startXMirror.toFloat(), startY, endXMirror.toFloat(), endY, fgPaint)
            }
        }
    }
}
